pkg/cmd/roachtest/cluster.go

```diff
@@ -1393,6 +1394,61 @@ func (c *clusterImpl) FetchDebugZip(
 	})
 }
 
+// FetchVMSpecs downloads the VM specs from the cluster using `roachprod get`.
+// The logs will be placed in the test's artifacts dir.
+func (c *clusterImpl) FetchVMSpecs(ctx context.Context, l *logger.Logger) error {
+	if c.IsLocal() {
+		return nil
+	}
+
+	l.Printf("fetching VM specs\n")
+	c.status("fetching VM specs")
+
+	vmSpecsFolder := filepath.Join(c.t.ArtifactsDir(), "vm_specs")
+	if err := os.MkdirAll(vmSpecsFolder, 0755); err != nil {
+		return err
+	}
+
+	// Don't hang forever if we can't fetch the specs.
+	return timeutil.RunWithTimeout(ctx, "fetch logs", 5*time.Minute, func(ctx context.Context) error {
+		cachedCluster, err := getCachedCluster(c.name)
+		if err != nil {
+			return err
+		}
+		providerToVMs := bucketVMsByProvider(cachedCluster)
+
+		for provider, vms := range providerToVMs {
+			p := vm.Providers[provider]
+			vmSpecs, err := p.GetVMSpecs(l, vms)
+			if err != nil {
+				l.Errorf("failed to get VM spec for provider %s: %s", provider, err)
+				continue
+			}
+			for _, vmSpec := range vmSpecs {
+				name, ok := vmSpec["name"].(string)
+				if !ok {
+					l.Printf("failed to create spec files for some VMs")
+					continue
+				}
+
+				dest := filepath.Join(vmSpecsFolder, name+".json")
+				specJSON, err := json.MarshalIndent(vmSpec, "", "  ")
+				if err != nil {
+					l.Printf("Failed to marshal JSON: %v\n", err)
+					continue
+				}
+
+				err = os.WriteFile(dest, specJSON, 0644)
+				if err != nil {
+					l.Printf("Failed to write spec to file for %s\n", name)
```

Comment:
Nit: `l.Errorf`

N.B. The rudimentary logger doesn't really have a notion of "log levels". Instead, the convention is to use `Errorf` for errors and warnings; the latter is typically prefixed with `warn: ` or `warning: `.

Reply:
Also, no need to write the newline in any of these logger calls.